Am/feat/more efficient integer list expansion

IceTDrinker opened the pull request:
<!-- Feel free to delete the template if the PR (bumping a version e.g.) does not fit the template -->
closes: _please link all relevant issues_

### PR content/description

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

tmontaigu commented on tfhe/src/shortint/ciphertext/zk.rs:
```diff
@@ -90,10 +90,57 @@ impl ProvenCompactCiphertextList {
+            } => match functions {
+                Some(functions) => {
+                    let functions_count = functions.len();
+                    let required_functions_count: usize = self
+                        .proved_lists
+                        .iter()
+                        .map(|list| list.0.ct_list.lwe_ciphertext_count().0)
+                        .sum();
```
I'm confused is it really one fn per ciphertext and not the possibility to run many functions for one ct ?

IceTDrinker replied:
sorry there are two levels of arrays

first one is per block, and can contain many functions

IceTDrinker replied:
this checks the first level, I'll rename to make things clearer

tmontaigu commented on tfhe/src/integer/ciphertext/compact_list.rs:
```diff
@@ -18,19 +19,91 @@ use tfhe_versionable::Versionize;
+        .generate_lookup_table(|x: u64| (x / carry_modulus) % message_modulus);
+    let msg_extract_bool = sks.key.generate_lookup_table(|x: u64| {
+        let tmp = x % message_modulus;
+        if tmp == 0 {
+            0u64
+        } else {
+            1u64
+        }
```
`u64::from(tmp != 0)` could also work

IceTDrinker replied:
indeed

tmontaigu commented on tfhe/src/integer/ciphertext/compact_list.rs:
```diff
@@ -324,9 +445,87 @@ impl ParameterSetConformant for CompactCiphertextList {
+}
+
+impl IntegerUnpackingToShortintCastingModeHelper {
+    pub fn new(ksk_view: KeySwitchingKeyView<'_>) -> Self {
```
Maybe this could  take a &ServerKey isntead and be used in `unpack_and_sanitize_message_and_carries` to share lut definitions

IceTDrinker replied:
indeed

IceTDrinker replied:
hmm actually, I don't pass luts to shortint as there are already hard coded functions in the casting (to make casting work) so I don't generate the LUTs but rather pass callbacks, what did you have in mind here ?

IceTDrinker replied:
ah right for the unpack to call into the function to generate the call backs

IceTDrinker commented:
Failure in integer test, this is not critical for the release

IceTDrinker commented:
fixed it, issue was a problem with output degree not being properly updated in a case we were interested in, other cases the degree was already properly handled, added an assert on degrees in the test case that failed, relaunched tests with approved label
